chanbackup/backupfile_test.go

```diff
@@ -0,0 +1,289 @@
+package chanbackup
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+func makeFakePackedMulti() (PackedMulti, error) {
+	newPackedMulti := make([]byte, 50)
+	if _, err := rand.Read(newPackedMulti[:]); err != nil {
+		return nil, fmt.Errorf("unable to make test backup: %v", err)
+	}
+
+	return PackedMulti(newPackedMulti), nil
+}
+
+func assertBackupMatches(t *testing.T, filePath string,
+	currentBackup PackedMulti) {
+
+	t.Helper()
+
+	packedBackup, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		t.Fatalf("unable to test file: %v", err)
+	}
+
+	if !bytes.Equal(packedBackup, currentBackup) {
+		t.Fatalf("backups don't match after first swap: "+
+			"expected %x got %x", packedBackup[:],
+			currentBackup)
+	}
+}
+
+func assertFileDeleted(t *testing.T, filePath string) {
+	t.Helper()
+
+	_, err := os.Stat(filePath)
+	if err == nil {
+		t.Fatalf("file %v still exists: ", filePath)
+	}
+}
+
+// TestUpdateAndSwap test that we're able to properly swap out old backups on
+// disk with new ones. Additionally, after a swap operation succeeds, then each
+// time we should only have the main backup file on disk, as the temporary file
+// has been removed.
+func TestUpdateAndSwap(t *testing.T) {
+	t.Parallel()
+
+	tempTestDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatalf("unable to make temp dir: %v", err)
+	}
+	defer os.Remove(tempTestDir)
+
+	testCases := []struct {
+		fileName     string
+		tempFileName string
+
+		oldTempExists bool
+
+		valid bool
+	}{
+		// Main file name is blank, should fail.
+		{
+			fileName: "",
+			valid:    false,
+		},
+
+		// Old temporary file still exists, should be removed. Only one
+		// file should remain.
+		{
+			fileName: filepath.Join(
+				tempTestDir, DefaultBackupFileName,
+			),
+			tempFileName: filepath.Join(
+				tempTestDir, DefaultTempBackupFileName,
+			),
+			oldTempExists: true,
+			valid:         true,
+		},
+
+		// Old temp doesn't exist, should swap out file, only a single
+		// file remains.
+		{
+			fileName: filepath.Join(
+				tempTestDir, DefaultBackupFileName,
+			),
+			tempFileName: filepath.Join(
+				tempTestDir, DefaultTempBackupFileName,
+			),
+			valid: true,
+		},
+	}
+	for i, testCase := range testCases {
+		// Ensure that all created files are removed at the end of the
+		// test case.
+		defer os.Remove(testCase.fileName)
```

Comment:
fwiw, these will return an error in the first case when the filenames are `""`. don't think it's an issue here though

Reply:
Yep, just a defer in a test.